roachtest: sqlsmith countingFileWriter nil pointer

[(roachtest).sqlsmith/setup=tpcc/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2280917&tab=buildLog) on [master@176c08c801a028d43d0d6f3f7044af078ff3f4f9](https://github.com/cockroachdb/cockroach/commits/176c08c801a028d43d0d6f3f7044af078ff3f4f9):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=no-mutations/run_1
	cluster.go:1651,context.go:135,cluster.go:1640,test_runner.go:823: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2280917-1600409617-05-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 3: 7954
		1: 8683
		2: dead
		4: 5198
		Error: UNCLASSIFIED_PROBLEM: 2: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1143
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:267
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1839
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:203
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1357
		Wraps: (3) 2: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpcc/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2280917&tab=artifacts#/sqlsmith/setup=tpcc/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x25e56f8]

goroutine 15222 [running]:
panic(0x4aa6dc0, 0x8b072c0)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc0132c9a88 sp=0xc0132c99f8 pc=0x7b1aa2
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:199
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:394 +0x3ec fp=0xc0132c9ab8 sp=0xc0132c9a88 pc=0x7c776c
github.com/cockroachdb/cockroach/pkg/col/colserde.(*countingWriter).Write(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/col/colserde/file.go:321
github.com/cockroachdb/cockroach/pkg/col/colserde.(*FileSerializer).Finish(0xc0022f5980, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/col/colserde/file.go:140 +0x158 fp=0xc0132c9b58 sp=0xc0132c9ab8 pc=0x25e56f8
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).writeFooterAndFlush(0xc000c01e60, 0x6197700, 0xc004c7c9c0, 0xa6f4b3, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:483 +0x36 fp=0xc0132c9bc8 sp=0xc0132c9b58 pc=0x25ea506
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close(0xc000c01e60, 0x6197700, 0xc004c7c9c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:398 +0x36d fp=0xc0132c9ca0 sp=0xc0132c9bc8 pc=0x25e9d5d
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*PartitionedDiskQueue).Close(0xc000a44160, 0x6197700, 0xc004c7c9c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/partitionedqueue.go:383 +0xec fp=0xc0132c9d00 sp=0xc0132c9ca0 pc=0x25ed1ec
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*externalSorter).Close(0xc006c34620, 0x6197700, 0xc004c7c9c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_sort.go:385 +0x131 fp=0xc0132c9d48 sp=0xc0132c9d00 pc=0x2636f71
github.com/cockroachdb/cockroach/pkg/sql/colexec.Closers.CloseAndLogOnErr(0xc001297b80, 0x2, 0x2, 0x6197700, 0xc004c7c9c0, 0xc01bae64a6, 0x7)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/operator.go:147 +0xe3 fp=0xc0132c9de0 sp=0xc0132c9d48 pc=0x2ad2273
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).close(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:103
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).runWithStream(0xc00261ae60, 0x6197700, 0xc004c7c9c0, 0x7f7ce398bbb8, 0xc000b32e00, 0xc000b32d60)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:342 +0x23d fp=0xc0132c9e58 sp=0xc0132c9de0 pc=0x31dd87d
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).Run(0xc00261ae60, 0x6197640, 0xc002036a80, 0x60f2960, 0xc0012948c0, 0x471dfafb00000001, 0xaf832faeb3493aa8, 0xc06651a082, 0x4, 0xc000b32d60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:180 +0x2f2 fp=0xc0132c9ef8 sp=0xc0132c9e58 pc=0x31dca12
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRemoteOutputStream.func1(0x6197640, 0xc0020ea800, 0xc00210e730)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:593 +0x123 fp=0xc0132c9f90 sp=0xc0132c9ef8 pc=0x31e6243
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc00b0dccc0, 0x6197640, 0xc0020ea800, 0xc00210e730, 0xc001b188c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1202 +0x44 fp=0xc0132c9fb8 sp=0xc0132c9f90 pc=0x31e8bf4
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0132c9fc0 sp=0xc0132c9fb8 pc=0x7e2271
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1201 +0x6f
```

--- a/pkg/sql/colcontainer/BUILD.bazel
+++ b/pkg/sql/colcontainer/BUILD.bazel
@@ -38,6 +38,8 @@ go_test(
         "//pkg/sql/colexecbase",
         "//pkg/sql/colmem",
         "//pkg/sql/execinfra",
+        "//pkg/sql/pgwire/pgcode",
+        "//pkg/sql/pgwire/pgerror",
         "//pkg/sql/types",
         "//pkg/storage/fs",
         "//pkg/testutils/colcontainerutils",

--- a/pkg/sql/colcontainer/diskqueue.go
+++ b/pkg/sql/colcontainer/diskqueue.go
@@ -495,9 +495,16 @@ func (d *diskQueue) resetWriters(f fs.File) error {
 	return d.serializer.Reset(d.writer)
 }
 
-func (d *diskQueue) writeFooterAndFlush(ctx context.Context) error {
-	err := d.serializer.Finish()
-	if err != nil {
+func (d *diskQueue) writeFooterAndFlush(ctx context.Context) (err error) {
+	defer func() {
+		if err != nil {
+			// If an error occurs, set the serializer to nil to avoid any future
+			// attempts to call writeFooterAndFlush during valid operation (e.g.
+			// calling Close after an error).
+			d.serializer = nil
+		}
+	}()
+	if err := d.serializer.Finish(); err != nil {
 		return err
 	}
 	written, err := d.writer.compressAndFlush()

--- a/pkg/sql/colcontainer/diskqueue_test.go
+++ b/pkg/sql/colcontainer/diskqueue_test.go
@@ -19,6 +19,9 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/col/coldatatestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/colcontainer"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/testutils/colcontainerutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
@@ -165,6 +168,35 @@ func TestDiskQueue(t *testing.T) {
 			}
 		}
 	}
+}
+
+func TestDiskQueueCloseOnErr(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
+	defer cleanup()
+
+	serverCfg := &execinfra.ServerConfig{}
+	serverCfg.TestingKnobs.MemoryLimitBytes = 1
+	diskMon := execinfra.NewLimitedMonitor(ctx, testDiskMonitor, serverCfg, t.Name())
+	defer diskMon.Stop(ctx)
+	diskAcc := diskMon.MakeBoundAccount()
+	defer diskAcc.Close(ctx)
+
+	typs := []*types.T{types.Int}
+	q, err := colcontainer.NewDiskQueue(ctx, typs, queueCfg, &diskAcc)
+	require.NoError(t, err)
+
+	b := coldata.NewMemBatch(typs, coldata.StandardColumnFactory)
+	b.SetLength(0)
+
+	err = q.Enqueue(ctx, b)
+	require.Error(t, err, "expected Enqueue to produce an error given a disk limit of one byte")
+	require.Equal(t, pgerror.GetPGCode(err), pgcode.DiskFull, "unexpected pg code")
+
+	// Now Close the queue, this should be successful.
+	require.NoError(t, q.Close(ctx))
 }
 
 // Flags for BenchmarkQueue.

